teps/0094-improve-task-resources-ux.md

```diff
@@ -0,0 +1,119 @@
+---
+status: proposed
+title: Improve UX of Task Resource Requirements
+creation-date: '2021-11-08'
+last-updated: '2021-11-08'
+authors:
+- '@lbernick'
+---
+
+# TEP-0094: Improve UX of Task Resource Requirements
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Currently, users can specify resource requirements in a `Task` definition, via the `Resources` field of each `Step`, `StepTemplate`, or `Sidecar`.
+However, the way users specify resource requirements doesn't reflect the way Tekton schedules `Task`s.
+In addition, these resource requirements cannot be parameterized or overridden at runtime.
+
+This TEP proposes updating the `Task` API to reflect how Tekton treats `Step` and `Sidecar` resource requirements,
+allowing these fields to be parameterized, and adding resource requirement configuration to `TaskRun`.
+
+## Motivation
+
+There are several issues affecting the user experience of specifying resource requirements for `Task`s:
+
+1. The way Tekton allows users to specify resource requirements is inconsistent with how `Task`s are actually scheduled,
+as described in [Issue: Improve UX of Step Resource Requests](https://github.com/tektoncd/pipeline/issues/2986).
+Specifically, all `Step` resource requests other than the maximum are ignored.
+
+2. We don't support parameterizing Step and Sidecar resource requirements. Even though we don't claim to support this,
+some users perceive the ability to parameterize some Task fields but not others [as a bug](https://github.com/tektoncd/community/pull/560#issuecomment-993348697).
+This feature was requested in [FR: Allow usage of variable replacement when defining resource limits and requests](https://github.com/tektoncd/pipeline/issues/4080).
+
+3. We don't support specifying resource requirements at runtime. Parameterizing resource requirements would not fully address this problem,
+as users should be able to specify resource requirements for Tasks they don't own, such as Catalog Tasks.
+This issue is described in more detail in [FR: Support specifying resource requests at TaskRun level](https://github.com/tektoncd/pipeline/issues/4326)
+
+This section elaborates on these issues.
+
+### Confusing Task Resource Requirements UX
+Under the hood, Tekton modifies the resource requests users specify to account for how the Tekton entrypoint runs each `Step` container.
+These modifications aren't transparent to users and have caused confusion.
```

Comment:
This is kind-of not true 😛. Starting in 0.28 (I think), tekton never modifies resource request users specified, **but** it will add some resource request/limits on anything that have not been specified by the user, if there is a (or several) `LimitRange` in the namespace.